chromadb/rate_limiting/admission_controller.py

```diff
@@ -0,0 +1,358 @@
+import threading
+import time
+import concurrent.futures
+import queue
+
+from typing import Optional, cast
+
+
+class Ticket:
+    def __init__(self, issued_by: Optional["AdmissionController"]) -> None:
+        self.acquisition_elapsed = 0.0
+        self.issued_by = issued_by
+
+    def release(self) -> None:
+        if self.issued_by:
+            self.issued_by.release()
+            self.issued_by = None
+
+
+class Waiter:
+    def __init__(self) -> None:
+        self.index = 0
+        self.woke_up = False
+
+    def notify(self) -> None:
+        self.woke_up = True
+
+
+class AdmissionController:
+    def admit_one(self) -> Optional[Ticket]:
+        raise NotImplementedError
+
+    def release(self) -> None:
+        raise NotImplementedError
+
+    def admitted(self) -> int:
+        raise NotImplementedError
+
+    def capacity(self) -> int:
+        raise NotImplementedError
+
+    def resize(self) -> int:
+        raise NotImplementedError
+
+    def stop(self) -> None:
+        raise NotImplementedError
+
+
+class AdmissionControllerImpl(AdmissionController):
+    FIFO_MODE = 0
+    LIFO_MODE = 1
+    DEFAULT_M = 0.005
+    DEFAULT_N = 0.1
+
+    def __init__(self, parallelism: int, M: float = DEFAULT_M, N: float = DEFAULT_N):
+        self.mtx = threading.Lock()
+        self.stopped = False
+        self.M = M
+        self.N = N
+        self.allowed = parallelism
+        self.admitted_count = 0
+        self.last_empty = time.time()
+        self.queue_mode = self.FIFO_MODE
+        self.waiters: list[Optional[Waiter]] = [None] * 64
```

Comment:
Why 64?

Reply:
Copied from my original code.  I suggest we make it something like 100_000 and get rid of the resize (and just outright fail if more than 100k requests back up).

Reply:
FWIW- The default queue in uvicorn is much smaller (4096 IIRC)

Reply:
Resolved offline—increasing queue size results in performance degradation and resizing works, so leaving as is